src/middleware/router.coffee

```diff
@@ -91,55 +98,74 @@ sendRequestToRoutes = (ctx, routes, next) ->
-            responseObj = JSON.parse response.body
-            ctx.mediatorResponse = responseObj
-            # then set koa response from responseObj.response
-            setKoaResponse ctx, responseObj.response
-          else
-            setKoaResponse ctx, response
-        .fail (reason) ->
-          # on failure
-          handleServerError ctx, reason
-      else
-        promise = buildNonPrimarySendRequestPromise ctx, route, options, path
-
-      promises.push promise
-
+      do (route) ->
+        path = getDestinationPath route, ctx.path
+        options =
+          hostname: route.host
+          port: route.port
+          path: path
+          method: ctx.request.method
+          headers: ctx.request.header
+          agent: false
+          rejectUnauthorized: true
+          key: keystore.key
+          cert: keystore.cert.data
+          secureProtocol: 'TLSv1_method'
+
+        if route.cert?
+          options.ca = keystore.ca.id(route.cert).data
+
+        if ctx.request.querystring
+          options.path += '?' + ctx.request.querystring
+
+        if options.headers && options.headers.authorization
+          delete options.headers.authorization
+
+        if route.username and route.password
+          options.auth = route.username + ":" + route.password
+
+        if options.headers && options.headers.host
+          delete options.headers.host
+
+        if route.primary
+          promise = sendRequest(ctx, route, options)
+          .then (response) ->
+            logger.info "executing primary route : #{route.name}"
+            if response.headers?['content-type']?.indexOf('application/json+openhim') > -1
+              # handle mediator reponse
+              responseObj = JSON.parse response.body
+              ctx.mediatorResponse = responseObj
+              # then set koa response from responseObj.response
+              setKoaResponse ctx, responseObj.response
+            else
+              setKoaResponse ctx, response
+          .then ->
+            logger.info "primary route completed"
+            next()
+
+          .fail (reason) ->
+            # on failure
+            handleServerError ctx, reason
+            next()
+        else
+          logger.info "executing non primary: #{route.name}"
+          promise = buildNonPrimarySendRequestPromise(ctx, route, options, path)
+          .then (routeObj) ->
+            logger.info "Storing non primary route responses #{route.name}"
+
+            try
+              messageStore.storeNonPrimaryResponse ctx, routeObj, ->
+                stats.nonPrimaryRouteRequestCount ctx, routeObj, ->
+                  stats.nonPrimaryRouteDurations ctx, routeObj, ->
+                    console.log 'sent durarions'
+            catch err
+              logger.error err
+
+        promises.push promise
     (Q.all promises).then ->
-      next()
+      messageStore.setFinalStatus ctx, ->
+        logger.info "all routes completed"
```

Comment:
Maybe add a little more info to this log message, like the transaction id it relates to.